Closes #3674 - Update stats_history.history_mysql_status_variables

jaredev opened the pull request:
Converted history table, history_mysql_status_variables to use an integer id called variable_id instead of string variable_name for scalability. Added lookup table to map these ids to their variable_name. Modified ProxySQL_Statistics to reflect these changes. Added tap test unit. Refer to commit messages and code documentation for more details. 5/6 tasks hopefully ready. 

Note the test unit compiles but I haven't verified it is operational. A such, should verify that there are no syntax errors and is operational.

jaredev committed:
Updated the SQLite history_mysql_status_variables create table statement to use variable_id instead of name for improved scalability. Added an index for the sql_status_variables table. Added lookup table create statement for history_mysql_status_variables_lookup table to refer the variable_id integer key to the variable_name string. Inserted the new table definition and execute code for the create index that were added. As the code has not yet received public release, there is no concern of keeping the previous table definition, and so the version number refers to v2.4.0.

jaredev committed:
Modified ProxySQL_Statistics to have two new member functions and two new private members. These are used to map the string variable_name to an integer variable_id to improve scalability. See documentation about their usage. Updated the MySQL_Threads_Hander_sets_v2 function to load the variable_name_id_map if it is empty for initial load from the lookup table. The prepared statement binds was changed to use the int64 variable_id retreived from the map.

jaredev committed:
Made the new member functions public. If these are only accessible internally by this class, or by the web interface plugin, consider making private and friend class to the plugin. As the extent of their usage is not yet known to me, making them public for now. Caution on adding new lookup entries unnecessarily.

jaredev committed:
In debug mode, the handlers for ProxySQL_Statistics are called even if the web interface plugin is not loaded. This way the code can be tested even if the plugin is not loaded. Added new member function ProxySQL_Statistics::knows_variable_name(), which is used to check if a variable_name is in the map. This is useful to do a check before get_variable_id_for_name() in instances where it is undesirable to load or create entries in the lookup table, while keeping direct access to the map private and mutex protected. In ProxySQL_Statistics::get_variable_id_for_name, wrapped the variable_id select code in a lambda so that it could be reused instead of calling sqlite3_last_insert_rowid(), as there were some thread safety concerns which potentially could produce incorrect ids. Changed the insert to use execute() instead of execute_statement() instead, sufficient and simpler. Added error in case the function could not find, load, or generate a new id.

jaredev committed:
Added the local_testing_datadir/ directory to gitignore, which is made during jenkins build script setup. Modified a tap Makefile to include curl path to resolve compilation error when compiling tap tests. Added tap test file 'test_admin_stats-t.cpp' for testing the ProxySQL_Statistics code changes, table definitions, and expected results.

pondix commented:
Automated message: PR pending admin approval for build testing

JavierJF commented on include/ProxySQL_Statistics.hpp:
```diff
@@ -131,6 +137,28 @@ class ProxySQL_Statistics {
+	 * @param variable_name The string variable name identifier
+	 * @return Integer variable id for the given variable name
+	 */
+	int64_t get_variable_id_for_name(std::string variable_name);
```
Here `variable_name` parameter could be `const &`.

JavierJF commented on include/ProxySQL_Statistics.hpp:
```diff
@@ -131,6 +137,28 @@ class ProxySQL_Statistics {
+	void load_variable_name_id_map_if_empty();
+
+	/** @return True if the given variable_name is registered in the variable_name_id_map. */
+	bool knows_variable_name(std::string variable_name) const;
```
Same here `variable_name` can be `const &`.

JavierJF commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -5081,7 +5081,11 @@ static void * admin_main_loop(void *arg)
 					SQLite3_result * resultset=MyHGM->SQL3_Get_ConnPool_Stats();
 					if (resultset) {
 						SQLite3_result * resultset2 = NULL;
+					#ifdef DEBUG
```
Sorry, maybe I'm out of loop here, I'm leaving this as a flag for later discussing if this is the path we want to follow.

jaredev replied:
This came from a discussion with Rene. The basic idea was to be able to test the metrics even if the web interface plugin isn't loaded in debug.

renecannao replied:
I confirm this decision. But we should document it, even a simple comment in the code.

jaredev replied:
Agreed. Adding comments.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+		int affected_rows;
+		char *error = NULL;
+		
+		auto select_var_id = [&]() -> int64_t {
```
I like lambdas, but in cases like this, I found them more easily to refactor (and read) in the future if they modify the less shared state as possible. Since all the state this lambda appears to need it in the very same context of this branch asides from the `variable_name` a simpler function signature could be: `[] (const std::string& variable_name) -> int64_t`. This way everything is packed in case we want to promote it to a regular function.

jaredev replied:
My reasoning was following along how similar code was written elsewhere in the module where these vars were shared by multiple selects. If done in similar style, this would produce near identical code but less redundant. It does also use statsdb_disk, which without reference capture would change the signature from [&]() -> int64_t to [](SQLite3DB *, const string& variable_name) -> int64_t. This would not be a simpler function signature. I'm not sure I follow on your readability argument, as the capture by reference can be treated as would any other scope in the same function.

JavierJF replied:
Yes, I see you point. I'm into everything that removes code duplication :100: . My suggestion was simply that most of the context the lambda is using, could be within the lambda scope, for me, the least shared state the lambda uses, the more simple to understand it's and it's simpler also to convert into a proper function in the future if you want. For example, promoting the `(SQLite3DB *, const string& variable_name) -> int64_t` function into a regular one, will be a matter of just copy pasting it outside de current function. So it's a matter of reuse. As you said, it can be treated as any other scope, but then, why having things in a upper scope that you can have in a lower one? Anyway, it's just a "protectionist way" of seeing it, you can ignore this as it was just a suggestion :)

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+		char *error = NULL;
+		
+		auto select_var_id = [&]() -> int64_t {
+			string var_id_query = "SELECT variable_id FROM history_mysql_status_variables_lookup WHERE variable_name=\"" + variable_name + "\"";
```
This string can be `const`.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+		if (variable_id < 0) {
+			// No match found, create a new record in the lookup table and then select the newly generated id
+			string insert_var_query = "INSERT INTO history_mysql_status_variables_lookup(variable_name) VALUES(\"" + variable_name + "\")";
+			if (statsdb_disk->execute(insert_var_query.c_str())) {
```
There is no check in case of `INSERT ...` failure.

jaredev replied:
The return of execute is bool, that's used for the check here. If there's an issue, this is handled in the last if/else block near the end of this function.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+			statsdb_disk->execute_statement(var_id_query.c_str(), &error , &cols , &affected_rows , &result);
+
+			if (error) {
+				proxy_error("SQLITE ERROR %s", error);
```
`proxy_error()` function and it's family doesn't add the line break ("\n"), so it's required to place it manually. Also, just for consistency if you could keep the format from other similar errors: `SQLITE CRITICAL error: %s. Shutting down.\n`.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+			// No match found, create a new record in the lookup table and then select the newly generated id
+			string insert_var_query = "INSERT INTO history_mysql_status_variables_lookup(variable_name) VALUES(\"" + variable_name + "\")";
+			if (statsdb_disk->execute(insert_var_query.c_str())) {
+				variable_id = select_var_id();
```
Since the last query executed was an `INSERT...` there is no need for performing a second query for getting the `id` from the previous query execution, same result can be achieved via `sqlite3_last_insert_rowid()` in the following way:

```
sqlite3* db = statsdb_disk->get_db();
variable_id = sqlite3_last_insert_rowid(db);
```
This way another query and potential disk access can be avoided.

jaredev replied:
That's how I wrote the code at first, when mentioned to Rene, he said this might causes issues from multi-threaded sqlite calls. I figured mutex lock should do, but he was saying that sqlite might still not lock. So that, if another insertion was done, if the timing was wrong, that the wrong id would be returned. I'd gladly change it back that way if this can be confirmed. That way as well, can remove the lambda above.

JavierJF replied:
Ha! Nice one... yes, indeed I think there is no guarantee that this is the only thread accessing `statsdb`, and if that is the case as René suggested, this would return the incorrect result. The lambda should stay for now.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+		if (variable_id > 0) {
+			variable_name_id_map[variable_name] = variable_id;
+		} else {
+			proxy_error("CRITIAL ERROR: Statistics could not find or generate variable_id for variable_name: %s", variable_name);
```
`variable_name` is of type `std::string` while `proxy_error` is expecting `const char *` this will result in runtime failure (also the line delimiter mentioned above is missing). Also there is a typo for 'CRITIAL' -> 'CRITICAL'.

JavierJF commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+	statsdb_disk->execute_statement(query.c_str(), &error , &cols , &affected_rows , &result);
+
+	if (error) {
+		proxy_error("SQLITE CRITICAL ERROR: %s", error);
```
Same as for the previous error comment for line '1107'.

JavierJF commented on test/tap/tests/test_admin_stats-t.cpp:
```diff
@@ -0,0 +1,180 @@
+
+	// If on a fresh install, wait long enough for the first run of stats to be created
+	// The lookup table will be empty until the first run!
+	sleep(new_stats_interval_sec * 2);
```
Maybe a little too much? `new_stats_interval_sec + 1` should be just fine. The valid intervals for being set are documented here: https://proxysql.com/documentation/global-variables/admin-variables/#admin-stats_mysql_connections. So something like the following will do the trick:

```
	// @TODO: run command to set the interval
	MYSQL_QUERY(proxysql_admin, "SET admin-stats_mysql_connections=5");
	MYSQL_QUERY(proxysql_admin, "LOAD ADMIN VARIABLES TO RUNTIME");
```

jaredev replied:
Thanks! Tentative, I wasn't sure what would be appropriate here to ensure enough time to populate initially. Updating...

JavierJF commented on test/tap/tests/test_admin_stats-t.cpp:
```diff
@@ -0,0 +1,180 @@
+
+	mysql_free_result(result);
+
+	ok(
```
For avoiding timing issues, you should wait at least `new_stats_interval_sec + 1` again before this second test executes, in the CI, it wont fail at this point, but in a freshly start `ProxySQL` and immediately launching this test. This _could_ not hold.

JavierJF commented on test/tap/tests/test_admin_stats-t.cpp:
```diff
@@ -0,0 +1,180 @@
+
+	for (int i = 0; i < result->row_count; i++) {
+		row = mysql_fetch_row(result);
+		rows_per_var_id.push_back(strtoll(row[0], nullptr, 10));
```
This is retrieving `variable_id` instead of `COUNT(*)`, if I have interpret this correctly, this should be: `strtoll(row[1])`.

jaredev replied:
Yes! Good catch. Resolving.

JavierJF commented on test/tap/tests/test_admin_stats-t.cpp:
```diff
@@ -0,0 +1,180 @@
+		"Each variable_id in the history table has the same number of rows."
+	);
+
+	// Test 5: Number of rows in history table increases after connections stat changes
```
The change itself should trigger the change of the table size. Waiting for another write due to a change in the interval should. If you want to verify that the interval change is holding, them, this final test could be changed into:

1. Changing the interval into a regular interval, and checking that the size holds the same for a while, maybe higher than the currently set interval.
2. Change the interval to a lower one, and check that the table size changed after the interval has expired.

For changing the interval, it can be done as previously described:

```
	MYSQL_QUERY(proxysql_admin, "SET admin-stats_mysql_connections=$NEW_INTERVAL");
	MYSQL_QUERY(proxysql_admin, "LOAD ADMIN VARIABLES TO RUNTIME");
```

jaredev replied:
I've reformulated the test, to be as such:

```
ok(
		(history_rows_before + distinct_var_ids_in_history) == history_rows_after,
		"Number of rows in history table increases correctly after insert interval change. Before: %lu After: %lu Difference should be: %lu",
		history_rows_before,
		history_rows_after,
		distinct_var_ids_in_history
	);
```

Using the lowest interval of 5 in earlier tests to speed up the  time to complete compared to beginning with a larger interval. Then, on the final test, increase it to 10, and wait. The number of records should match this condition above. For, if the interval was not updated correctly and remained at 5 these would not be equal, but would be twice as many records inserted.

JavierJF commented on test/tap/tests/test_admin_stats-t.cpp:
```diff
@@ -0,0 +1,180 @@
+/**
+ * @file test_admin_stats-t.cpp
+ * @brief This tests the Statistics module and its lookup code and tables.
```
A @details section detailing which each of the tests performed is aim to verify would be nice, this way there is no need to reading the test code for knowing everything it aims to check.

jaredev replied:
OK. I was just going off of example in test_admin_prometheus_metrics_dump-t... as a starting point. I see now looking at another test what you mean. I'll add that in there.

jaredev committed:
Const strings where appropriate and improved test for statistics

Changed the get_variable_id_for_name, and knows_variable_name to use const string &. Changed immutable strings to to be const as well. Added details comment to test_admin_stats-t.cpp test. In the test code, improved documentation, set a valid interval for metrics insertion, and added the commands to set the interval. Changed the sleep time to be the inteval plus a second should be long enough to let new metrics data be added. Put additional sleep before test 2. Fixed incorrect column index usage in test 4 result set. Reformulated test 5, to check the number of records inserted after new interval is correct.

jaredev committed:
Fixed error messages

Added new lines for the proxy_error() calls, and following similar convention as others, fixed a typo, and type conversion.

renecannao commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -644,7 +644,11 @@ void ProxySQL_Statistics::system_memory_sets() {
 void ProxySQL_Statistics::MyHGM_Handler_sets(SQLite3_result *resultset1, SQLite3_result *resultset2) {
 	MyHGM_Handler_sets_v1(resultset1);
+#ifdef DEBUG
+	if (resultset2) {
```
We should add a comment that mentions we enable this feature in debug mode even if web interface is not enabled, for debugging and testing purpose.
Without comment we will forget (especially me).

renecannao commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -789,7 +793,11 @@ void ProxySQL_Statistics::MyHGM_Handler_sets_v1(SQLite3_result *resultset) {
 void ProxySQL_Statistics::MySQL_Threads_Handler_sets(SQLite3_result *resultset) {
 	MySQL_Threads_Handler_sets_v1(resultset);
+#ifdef DEBUG
+	if (true) {
```
We should add a comment that mentions we enable this feature in debug mode even if web interface is not enabled, for debugging and testing purpose.
Without comment we will forget (especially me).

renecannao commented on lib/ProxySQL_Statistics.cpp:
```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
+				if (result->rows_count > 0) {
+					// matching variable_id for variable_name in lookup table
+					SQLite3_row *r = result->rows[0];
+					return strtoll(r->fields[0], NULL, 10);
```
There is a memory leak here.
`result` is not deleted before returning

jaredev replied:
Good catch

renecannao commented on include/ProxySQL_Statistics.hpp:
```diff
@@ -5,22 +5,27 @@
 #define STATSDB_SQLITE_TABLE_MYSQL_CONNECTIONS STATSDB_SQLITE_TABLE_MYSQL_CONNECTIONS_V2_0
 
-#define STATSDB_SQLITE_TABLE_HISTORY_MYSQL_STATUS_VARIABLES_V2_0_10 "CREATE TABLE history_mysql_status_variables (timestamp INT NOT NULL , variable_name VARCHAR NOT NULL , variable_value VARCHAR NOT NULL , PRIMARY KEY (timestamp, variable_name))"
+#define STATSDB_SQLITE_TABLE_HISTORY_MYSQL_STATUS_VARIABLES_V2_4_0 "CREATE TABLE history_mysql_status_variables (timestamp INT NOT NULL , variable_id VARCHAR NOT NULL , variable_value VARCHAR NOT NULL , PRIMARY KEY (timestamp, variable_id))"
```
`variable_id` should be an `INT NOT NULL` .
The rest of the code treats it as an integer, so I assume this was just a typo

jaredev replied:
Correct. Easy fix

renecannao committed:
Merge branch 'v2.x' into v2.x-3674

renecannao commented:
add to whitelist

jaredev committed:
Fixed typo, memory leak, and documentation.

* Fixed typo for variable_id in history_mysql_status_variables table definition to be INT NOT NULL as intended.

* Added comments before debug ifdefs to explain their purpose

* Optimized get_variable_id_for_name by performing one less search by using iterator insteaed of knows_variable_name() call.

* Fixed memory leak by deleting resultset before return in lambda.

jaredev committed:
Merge branch 'v2.x-3674' of https://github.com/jaredev/proxysql into v2.x-3674
